user/TASK/gimbal_task/gimbal_task.c

```diff
@@ -89,29 +111,58 @@ static void initialization(Gimbal_t *gimbal_ptr){
     gimbal_ptr->launcher = get_launcher_pointer();
     fp32 pid_constants_yaw[3] = {pid_kp_yaw, pid_ki_yaw, pid_kd_yaw};
     fp32 pid_constants_pitch[3] = {pid_kp_pitch, pid_ki_pitch, pid_kd_pitch};
-
+    
+    gimbal_ptr->yaw_setpoint[0] = 0.0;
+    gimbal_ptr->yaw_setpoint[1] = -1.0;
+    gimbal_ptr->yaw_position[0] = 0.0;
+    gimbal_ptr->yaw_position[1] = -1.0;
+    gimbal_ptr->yaw_error = 0.0;
+    
     PID_Init(&(gimbal_ptr->pitch_motor.pid_controller), PID_POSITION, pid_constants_pitch, max_out_pitch, max_i_term_out_pitch);
     PID_Init(&(gimbal_ptr->yaw_motor.pid_controller), PID_POSITION, pid_constants_yaw, max_out_yaw, max_i_term_out_yaw);
     
     gimbal_ptr->rc_update = get_remote_control_point();
     
-    gimbal_ptr->pitch_motor.pos_set = 5500;
+    gimbal_ptr->pitch_motor.pos_set = 3000;
     gimbal_ptr->yaw_motor.pos_set = 6000;
+    // TODO: find better bounds on Pitch
 }
 
 /** 
  * @brief  Update encoder positions
  * @param  None
  * @retval None
  */
-static void get_new_data(Gimbal_t *gimbal_data){
-        // Get CAN received data 
-        gimbal_data->pitch_motor.pos_read = gimbal_data->pitch_motor.motor_feedback->ecd;
-        gimbal_data->pitch_motor.speed_read = gimbal_data->pitch_motor.motor_feedback->speed_rpm;
+static void get_new_data(Gimbal_t *gimbal_data){  
+     
+    // Get CAN received data 
+    gimbal_data->pitch_motor.pos_read = gimbal_data->pitch_motor.motor_feedback->ecd;
+    gimbal_data->pitch_motor.speed_read = gimbal_data->pitch_motor.motor_feedback->speed_rpm;
 
-        gimbal_data->yaw_motor.pos_read = gimbal_data->yaw_motor.motor_feedback->ecd;
-        gimbal_data->yaw_motor.speed_read = gimbal_data->yaw_motor.motor_feedback->speed_rpm;
+    gimbal_data->yaw_motor.pos_read = gimbal_data->yaw_motor.motor_feedback->ecd;
+    gimbal_data->yaw_motor.speed_read = gimbal_data->yaw_motor.motor_feedback->speed_rpm;
     
+    fill_complex_equivalent(gimbal_data->yaw_position, gimbal_data->yaw_motor.pos_read);
+}
+
+
+static void fill_complex_equivalent(fp32 position[2], uint16_t ecd_value){
```

Comment:
Should functions like this, operations and things that don't appear in the main control loop, all be below the the main functions? ie; the file should look like:

main_loop(){
a()
b()
}

a(){
    op1()
}

b(){
}

op1(){
}

Reply:
They do don't they??

Reply:
Currently the layout is:

gimbal_task()
initialization()
get_new_data()
**fill_complex_equivalent()**
**multiply_complex_a_by_b()**
**make_unit_length()**
update_setpoints()
**a_dot_b()**
**length_of_a_cross_b()**
**get_error_sign()**
increment_PID()
get_vision_signal()
send_to_uart()

I have marked in bold what I consider to be operations. I would put them all somewhere at the bottom of the file so that when you are reading the code, the interesting functions appear first, and in the order in which they appear in gimbal_task(). I think its easier to read that way.

Afterthought: If we end up with lots of little operation functions, it may be worth it to put them in user_lib.c or something like that.